File: leetcode_4/src/leetcode_4/guess_number.rs
/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * unsafe fn guess(num: i32) -> i32 {}
 */


#[allow(dead_code)]
unsafe fn guess(num: i32) -> i32 {
    num
}
struct Solution;

impl Solution {
    #[allow(dead_code)]
    unsafe fn guess_number(n: i32) -> i32 {
        let (mut left, mut right) = (1,n+1);
        loop {
            let mid = left + (right-left)/2;
            match guess(mid) {
                -1 => {
                    right = mid;
                },
                0 => {
                    return mid;
                },
                1 => {
                    left = mid+1;
                },
                _ => {
                    // never arrived
                }
            }
        }

    }
}

File: leetcode_cpp/Leetcode75/Binary Search/guessNumber.cpp
#include <bits/stdc++.h>
using namespace std;
auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

int guess(int num);

class Solution {
  public:
    static int guessNumber(int n) {
        unsigned int l = 1;
        unsigned int r = n;
        r++;
        while (l < r) {
            auto m = ((r - l) / 2) + l;
            auto g = guess(m);
            if (g == -1) {
                r = m;
            } else if (g == 1) {
                l = m + 1;
            } else {
                return m;
            }
        }
        std::unreachable();
    }
};